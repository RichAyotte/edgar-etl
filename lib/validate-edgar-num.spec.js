/**
 * @overview    Validate EDGAR numbers spec
 * @author      Richard Ayotte
 * @copyright   Copyright Â© 2019 Richard Ayotte
 * @license     MIT License
 */

'use strict'

const fs = require('fs')
const {Writable} = require('stream')
const parseEdgarNum = require('./parse-edgar-num')
const validateEdgarNum = require('./validate-edgar-num')

const getValidationErrors = async () => new Promise((resolve, reject) => {
	const validationErrors = []
	fs.createReadStream(`./sample-data/num-errors.txt`)
	.pipe(parseEdgarNum)
	.pipe(validateEdgarNum)
	.on('validation-error', error => {
		validationErrors.push(error)
	})
	.pipe(new Writable({
		objectMode: true
		, write(data, _, done) {
			done()
		}
	}))
	.on('finish', () => {
		resolve(validationErrors)
	})
	.on('error', error => {
		reject(error)
	})
	.on('skip', error => {
		reject(error)
	})
})


describe('EDGAR numbers validation', () => {
	it(`Emits validation-error`, async () => {
		expect.assertions(1)
		const validationErrors = await getValidationErrors()
		expect(validationErrors).toMatchSnapshot()
	})
})
