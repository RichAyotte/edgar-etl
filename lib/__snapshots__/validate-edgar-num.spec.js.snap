// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EDGAR numbers validation Emits validation-error 1`] = `
Array [
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20090331,
      "footnote": "",
      "qtrs": 1,
      "tag": "CommonStockDividendsPerShareDeclared",
      "uom": "pure",
      "value": 0.55,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20080331,
      "footnote": "",
      "qtrs": 1,
      "tag": "CommonStockDividendsPerShareDeclared",
      "uom": "pure",
      "value": 0.55,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20090331,
      "footnote": "",
      "qtrs": 1,
      "tag": "EarningsPerShareBasic",
      "uom": "pure",
      "value": 0.39,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20080331,
      "footnote": "",
      "qtrs": 1,
      "tag": "EarningsPerShareBasic",
      "uom": "pure",
      "value": 0.91,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20090331,
      "footnote": "",
      "qtrs": 1,
      "tag": "EarningsPerShareDiluted",
      "uom": "pure",
      "value": 0.39,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20080331,
      "footnote": "",
      "qtrs": 1,
      "tag": "EarningsPerShareDiluted",
      "uom": "pure",
      "value": 0.9,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
  Object {
    "data": Object {
      "adsh": "0001031296-09-000011",
      "coreg": "",
      "ddate": 20090331,
      "footnote": "",
      "qtrs": 1,
      "tag": "EarningsPerShareBasic",
      "uom": "pure",
      "value": 0.39,
      "version": "us-gaap/2008",
    },
    "error": [TypeError: expecting uom to be "USD" or "shares", got String "pure"],
  },
]
`;
